target_chains/ton/contracts/contracts/Wormhole.fc

```diff
@@ -0,0 +1,237 @@
+}
+
+(int, int, int, int, int, slice, int, int, slice, int) parse_and_verify_wormhole_vm(slice in_msg_body) impure {
+    ;; Parse VM fields
+    int version = in_msg_body~load_uint(8);
+    throw_unless(ERROR_INVALID_VERSION, version == 1);
+    int vm_guardian_set_index = in_msg_body~load_uint(32);
+    ;; Verify and check if guardian set is valid
+    (int expiration_time, cell keys, int keys_count) = get_guardian_set_internal(vm_guardian_set_index);
+    throw_if(ERROR_INVALID_GUARDIAN_SET_KEYS_LENGTH, cell_null?(keys));
+    throw_unless(ERROR_INVALID_GUARDIAN_SET,
+        (guardian_set_index == vm_guardian_set_index) &
+        ((expiration_time == 0) | (expiration_time > now()))
+    );
+    int signers_length = in_msg_body~load_uint(8);
+    ;; Calculate hash_index in bytes (66 bytes per signature: 1 (guardianIndex) + 32 (r) + 32 (s) + 1 (v))
+    int hash_index = signers_length * 66;
+
+    ;; Load signatures
+    (cell signatures, slice remaining_body) = read_and_store_large_data(in_msg_body, hash_index * 8);
+    in_msg_body = remaining_body;
+
+    ;; Calculate total body length across all references
+    int body_length = 0;
+    int continue? = -1; ;; -1 is true
+    do {
+        body_length += remaining_body.slice_bits();
+        if (remaining_body.slice_refs_empty?()) {
+            continue? = 0;
+        } else {
+            remaining_body = remaining_body~load_ref().begin_parse();
+        }
+    } until (~ continue?);
+
+    ;; Load body
+    (cell body_cell, _) = read_and_store_large_data(in_msg_body, body_length);
+
+    int hash = hash_vm_body(body_cell.begin_parse());
+    ;; Verify signatures
+    verify_signatures(hash, signatures.begin_parse(), signers_length, keys, keys_count);
+
+    slice body_slice = body_cell.begin_parse();
+    int timestamp = body_slice~load_uint(32);
+    int nonce = body_slice~load_uint(32);
+    int emitter_chain_id = body_slice~load_uint(16);
+    slice emitter_address = body_slice~load_bits(256);
+    int sequence = body_slice~load_uint(64);
+    int consistency_level = body_slice~load_uint(8);
+    slice payload = body_slice;
+
+    return (
+        version,
+        vm_guardian_set_index,
+        timestamp,
+        nonce,
+        emitter_chain_id,
+        emitter_address,
+        sequence,
+        consistency_level,
+        payload,
+        hash
+    );
+}
+
+(int, int, int, cell, int) parse_encoded_upgrade(int current_guardian_set_index, slice payload) impure {
+    slice module = payload~load_bits(32);
+    int module_int = module~load_uint(32);
+    throw_unless(ERROR_INVALID_MODULE, module_int == UPGRADE_MODULE);
+
+    int action = payload~load_uint(8);
+    throw_unless(ERROR_INVALID_GOVERNANCE_ACTION, action == 2);
+
+    int chain = payload~load_uint(16);
+    int new_guardian_set_index = payload~load_uint(32);
+    throw_unless(ERROR_NEW_GUARDIAN_SET_INDEX_IS_INVALID, new_guardian_set_index > current_guardian_set_index);
```

Comment:
i think the condition for this is new_guardian_set_index == current_guardian_set_index + 1 in other places